packages/cli/src/lib/setup/setupSetup.ts

```diff
@@ -1024,6 +1022,214 @@ Please DO NOT copy files manually into ioBroker storage directories!`
+        } catch (e) {
+            console.error(`Cannot clean up garbage objects: ${e.message}`);
+        }
+
+        console.log('Clean up leftover adapters ...');
+        try {
+            await this._cleanupLeftoverAdapters();
+        } catch (e) {
+            console.error(`Cannot clean up leftover adapters: ${e.message}`);
+        }
+
+        console.log('Clean up objects and states with forbidden characters ...');
+        try {
+            await this._cleanupForbiddenIds();
+        } catch (e) {
+            console.error(`Cannot clean up objects and states with forbidden IDs: ${e.message}`);
+        }
+    }
+
+    /**
+     * Cleanup adapter objects from already removed adapters, which are still there due to errors or past bugs
+     */
+    private async _cleanupLeftoverAdapters(): Promise<void> {
+        if (!this.objects) {
+            throw new Error('Objects not set up, call setupObjects first');
+        }
+
+        const hostsView = await this.objects.getObjectViewAsync('system', 'host', {
+            startkey: SYSTEM_HOST_PREFIX,
+            endkey: `${SYSTEM_HOST_PREFIX}\u9999`
+        });
+
+        const hostIds = hostsView.rows.map(row => row.id);
+
+        for (const hostId of hostIds) {
+            const hasAdapters = await this.objects.objectExists(`${hostId}.adapters`);
+
+            if (!hasAdapters) {
+                console.log(
+                    `Skipping cleanup leftover adapters, because host "${hostId}" is not yet migrated to a supporting controller version`
+                );
+                return;
+            }
+        }
+
+        const adaptersViewPerHost = await this.objects.getObjectViewAsync('system', 'adapter', {
+            startkey: SYSTEM_HOST_PREFIX,
+            endkey: `${SYSTEM_HOST_PREFIX}\u9999`
+        });
+
+        const installedAdapterNames = adaptersViewPerHost.rows.map(row => row.value.common.name);
+
+        const adaptersView = await this.objects.getObjectViewAsync('system', 'adapter', {
+            startkey: SYSTEM_ADAPTER_PREFIX,
+            endkey: `${SYSTEM_ADAPTER_PREFIX}\u9999`
+        });
+
+        for (const row of adaptersView.rows) {
+            if (!installedAdapterNames.includes(row.value.common.name)) {
+                await this.objects.delObject(row.id);
+                console.log(`Cleaned up leftover adapter object "${row.id}"`);
+            }
+        }
+    }
+
+    /**
+     * Cleanup all states and objects which contain forbidden chars in their id
+     */
+    private async _cleanupForbiddenIds(): Promise<void> {
+        if (!this.objects) {
+            throw new Error('Objects not set up, call setupObjects first');
+        }
+
+        if (!this.states) {
+            throw new Error('States not set up, call setupObjects first');
+        }
+
+        const objIds = await this.objects.getKeys('*');
+
+        if (objIds) {
+            for (const id of objIds) {
+                if (FORBIDDEN_CHARS.test(id)) {
```

Comment:
wikl that work and not throw? Do you might need the special mpde as option for that that ignores these character errors?

Reply:
no on db level there is no check